crates/turbopack-core/src/source_map.rs

```diff
@@ -0,0 +1,303 @@
+                name: t.get_name().map(|n| n.to_string()),
+            })
+        } else {
+            Token::Synthetic(SyntheticToken {
+                generated_line: t.get_dst_line() as usize,
+                generated_column: t.get_dst_col() as usize,
+            })
+        }
+    }
+}
+
+impl SourceMap {
+    /// Encoding a SourceMap stringifies it into JSON.
+    #[async_recursion]
+    async fn encode<W: Write + Send>(&self, w: &mut W) -> Result<()> {
+        match self {
+            SourceMap::Regular(r) => r.0.to_writer(w)?,
+
+            SourceMap::Sectioned(s) => {
+                // My kingdom for a decent dedent macro with interpolation!
+                write!(
+                    w,
+                    r#"{{
+  "version": 3,
+  "sections": ["#
+                )?;
+
+                let sections = s
+                    .sections
+                    .iter()
+                    .map(async move |s| Ok((s.offset, s.map.await?)))
+                    .try_join()
+                    .await?;
+
+                let mut first_section = true;
+                for (offset, map) in sections {
+                    if !first_section {
+                        write!(w, ",")?;
+                    }
+                    first_section = false;
+
+                    write!(
+                        w,
+                        r#"
+    {{"offset": {{"line": {}, "column": {}}}, "map": "#,
+                        offset.line, offset.column,
+                    )?;
+
+                    map.encode(w).await?;
+                    write!(w, r#"}}"#)?;
+                }
+
+                write!(
+                    w,
+                    r#"]
+}}"#
+                )?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl SourceMapVc {
+    /// Creates a new SourceMap::Regular Vc out of a sourcemap::SourceMap
+    /// ("CrateMap") instance.
+    pub fn new_regular(map: CrateMap) -> Self {
+        SourceMap::Regular(RegularSourceMap::new(map)).cell()
+    }
+
+    /// Creates a new SourceMap::Sectioned Vc out of a collection of source map
+    /// sections.
+    pub fn new_sectioned(sections: Vec<SourceMapSection>) -> Self {
+        SourceMap::Sectioned(SectionedSourceMap::new(sections)).cell()
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl SourceMapVc {
+    /// Stringifies the source map into JSON bytes.
+    #[turbo_tasks::function]
+    pub async fn to_bytes(self) -> Result<BytesVc> {
+        let mut bytes = vec![];
+        self.await?.encode(&mut bytes).await?;
+        Ok(BytesVc::cell(bytes))
+    }
+
+    /// Traces a generated line/column into an mapping token representing either
+    /// synthetic code or user-authored original code.
+    #[turbo_tasks::function]
+    pub async fn lookup_token(self, line: usize, column: usize) -> Result<OptionTokenVc> {
+        let token = match &*self.await? {
+            SourceMap::Regular(map) => {
+                match map.lookup_token(line as u32, column as u32) {
```

Comment:
What's the reasoning for storing line and column as `usize`?

Reply:
`map` here is a `CrateMap`, which expects `u32`. Our internal `SourcePos` is used all over the place, including calling code rust stdlib code, and everything expects `usize` there.

We could probably save bytes by storing everything as a `u32`, but it'll require casting in a lot of existing code.